timemachine/fe/absolute_hydration.py

```diff
@@ -0,0 +1,142 @@
+    ligand_xvs = enhanced.sample_from_log_weights(ligand_samples, ligand_log_weights, size=num_samples)
+
+    # sample uniformly with replacement
+    solvent_choice_idxs = np.random.choice(np.arange(len(solvent_samples)), size=num_samples)
+
+    all_xvbs = []
+    for i, choice_idx in enumerate(solvent_choice_idxs):
+
+        # solvent + noninteracting ligand
+        noninteracting_xvb = solvent_samples[choice_idx]
+
+        # vacuum ligand
+        ligand_x, ligand_v = ligand_xvs[i]
+
+        # concatenate solvent
+        combined_x = np.concatenate([noninteracting_xvb.coords[solvent_idxs], ligand_x], axis=0)
+        combined_v = np.concatenate([noninteracting_xvb.velocities[solvent_idxs], ligand_v], axis=0)
+
+        combined_xvb = CoordsVelBox(combined_x, combined_v, noninteracting_xvb.box)
+
+        all_xvbs.append(combined_xvb)
+    return all_xvbs
+
+
+def setup_absolute_hydration_with_endpoint_samples(mol, temperature=300.0, pressure=1.0, n_steps=1000, seed=2022):
+    """Generate samples from the equilibrium distribution at lambda=1
+
+    Return:
+    * reduced_potential_fxn
+    * npt_mover
+    * initial_samples from lam = 1
+    """
+    if type(seed) != int:
+        seed = np.random.randint(1000)
+        print(f"setting seed randomly to {seed}")
+    else:
+        print(f"setting seed to {seed}")
+
+    np.random.seed(seed)
+
+    # set up potentials
+    ff = Forcefield.load_from_file("smirnoff_1_1_0_ccc.py")
+    potentials, params, masses, _, _ = enhanced.get_solvent_phase_system(mol, ff)
+    potential_fxn = construct_potential(potentials, params)
+
+    kBT = BOLTZ * temperature
+
+    def reduced_potential_fxn(xvb, lam):
+        return potential_fxn(xvb, lam) / kBT
+
+    bind_potentials(potentials, params)
+
+    # set up npt mover
+    npt_mover = moves.NPTMove(potentials, None, masses, temperature, pressure, n_steps, seed)
+
+    # combine solvent and ligand samples
+    solvent_xvbs, ligand_samples, ligand_log_weights = enhanced.load_or_pregenerate_samples(
+        mol, ff, seed, temperature=temperature, pressure=pressure
+    )
+    n_endstate_samples = 5000  # TODO: expose this parameter?
+    num_ligand_atoms = mol.GetNumAtoms()
+    all_xvbs = generate_endstate_samples(
+        n_endstate_samples, solvent_xvbs, ligand_samples, ligand_log_weights, num_ligand_atoms
+    )
+
+    return reduced_potential_fxn, npt_mover, all_xvbs
+
+
+def set_up_ahfe_system_for_smc(mol, n_walkers, n_windows, n_md_steps, resample_thresh, seed=2022):
+    """define initial samples, lambdas schedule, propagate fxn, log_prob fxn, resample fxn"""
+    reduced_potential, mover, initial_samples = setup_absolute_hydration_with_endpoint_samples(
+        mol, n_steps=n_md_steps, seed=seed
+    )
+
+    sample_inds = np.random.choice(np.arange(len(initial_samples)), size=n_walkers)
+    samples = [initial_samples[i] for i in sample_inds]
+
+    lambdas = construct_pre_optimized_absolute_lambda_schedule_solvent(n_windows)[::-1]
```

Comment:
worth commenting on why this is reversed?

Reply:
comment added in https://github.com/proteneer/timemachine/pull/718/commits/18ccc6997b339fed02220b4194623ca68b090b64